model-optimizer/extensions/back/MarkNodesWithShapeValues.py

```diff
@@ -52,27 +53,48 @@ def get_operations_with_shape_inputs():
+        shape_accepting_nodes = []
         for node in graph.get_op_nodes():
-            if node.soft_get('type') in shape_input_ops_map:
-                nodes_with_shape_inputs.append(node)
+            if node.soft_get('type') in shape_accepting_ops:
+                shape_accepting_nodes.append(node)
+        return shape_accepting_nodes
 
-        start_nodes = []
+    @staticmethod
+    def get_sources_for_nodes(nodes_with_shape_inputs: List[Node], filter_condition: Callable) -> List[Node]:
+        shape_accepting_ops = MarkNodesWithShapeValues.get_shape_accepting_ops()
+        sources = []
         for node in nodes_with_shape_inputs:
-            start_nodes.extend(
-                [node.in_port(port_idx).get_source().node for port_idx in shape_input_ops_map[node.type] if
-                 node.is_in_port_connected(port_idx)])
+            for port_idx in shape_accepting_ops[node.soft_get('type')]:
+                if not node.is_in_port_connected(port_idx):
+                    continue
 
-        condition = lambda node: node.soft_get('type') != 'ShapeOf'
-        nodes_with_shape_values = MarkSubGraphsWithCorrectLayout.bfs(start_nodes, set(), condition, forward=False)
-        for node in nodes_with_shape_values:
+                source_node = node.in_port(port_idx).get_source().node
+                if not filter_condition(source_node):
+                    continue
+
+                sources.append(source_node)
+        return sources
+
+    @staticmethod
+    def mark_nodes(shape_returning_nodes: List[Node]):
+        for node in shape_returning_nodes:
             node['returns_shape_value'] = True
             if node.soft_get('type') == 'Const':
                 if node.value.dtype == np.float32:
                     node.out_node(0)['correct_data_type'] = True
                 elif node.value.dtype in [np.float16, np.float64]:
-                    log.debug('Const nodes {} with shape values have {} type'.format(node.soft_get('name', node.id),
-                                                                                     node.value.dtype))
+                    log.error("Const node '{}' returns shape values of '{}' type but it must be integer or float32. "
+                              'During Elementwise type inference will attempt to cast to float32'.
+                              format(node.soft_get('name', node.id), node.value.dtype), extra={'is_warning': True})
+
+    def find_and_replace_pattern(self, graph: Graph):
+        shape_accepting_nodes = self.get_nodes_with_shape_inputs(graph)
+
+        condition = lambda node: node.soft_get('type') != 'ShapeOf'
+        sources_of_shape_accepting_nodes = self.get_sources_for_nodes(shape_accepting_nodes, condition)
+        shape_returning_nodes = MarkSubGraphsWithCorrectLayout.bfs(sources_of_shape_accepting_nodes, set(),
+                                                                   condition, forward=False)
```

Comment:
Also, we can consider to move `bfs` method move into utils

Reply:
that's a good idea! I also thought the same, but unfortunately there are several BFS search functions (AFAIK more than 4) scattered around MO, some of them are already in utils some not
https://github.com/openvinotoolkit/openvino/blob/master/model-optimizer/mo/utils/graph.py#L15

When was writing MarkNodesWithShapeValues I was not able to reuse bfs from utils, but only from MarkSubGraphsWithCorrectLayout 

If we want to mode this bfs function then other transformations not related to the current tickets should be changed as well, e.g. MarkSubGraphsWithCorrectLayout but those transformation do not have unit-tests so probability to make an error increases.

I suggest to revise BFS use in whole MO, but in the next release since we passed FF such changes now can be dangerous. @lazarevevgeny how do you think?

Reply:
Unfortunately, it is not possible to implement common bfs function which will fulfil everybody's needs. So let's keep using existing functions in the place where it is used.